python/ray/serve/_private/deployment_scheduler.py

```diff
@@ -294,6 +358,95 @@ def _get_node_to_running_replicas(
 
         return res
 
+    def _get_available_resources_per_node(self) -> Dict[str, Resources]:
+        """Gets current available resources per node.
+
+        This returns a conservative view of the available resources
+        currently in the cluster. It returns the minimum of:
+
+        1. The available resources per node fetched and cached from the
+           GCS every control loop.
+        2. The remaining resources left over on each node after
+           subtracting the resources taken up by running (already
+           scheduled by core) and launching (to-be-scheduled and soft
+           targeting that node) replicas.
+
+        Note that (1) may not be accurate because it uses cached info
+        and there is a delay from fetching data from GCS, and (2) may
+        not be accurate because there can be other actors (not replicas)
+        running in the cluster, and launching replicas may not end up on
+        the node we're targeting. So the information returned from this
+        method is only best effort.
+        """
+
+        available_resources = (
+            self._cluster_node_info_cache.get_available_resources_per_node()
+        )
+        total_resources = self._cluster_node_info_cache.get_total_resources_per_node()
+
+        gcs_info = {node_id: Resources(r) for node_id, r in available_resources.items()}
+
+        # Manually calculate available resources per node by subtracting
+        # launching and running replicas from total resources
+        available_minus_replicas = {
+            node_id: Resources(resources)
+            for node_id, resources in total_resources.items()
+        }
+        for deployment_id, replicas in self._launching_replicas.items():
+            for info in replicas.values():
+                if info.target_node_id:
+                    available_minus_replicas[info.target_node_id] -= self._deployments[
+                        deployment_id
+                    ].required_resources
+        for deployment_id, replicas in self._running_replicas.items():
+            for node_id in replicas.values():
+                available_minus_replicas[node_id] -= self._deployments[
+                    deployment_id
+                ].required_resources
+
+        def custom_min(a: Resources, b: Resources):
+            keys = set(a.keys()) | set(b.keys())
+            res = Resources()
+            for key in keys:
+                res[key] = min(a.get(key), b.get(key))
+            return res
+
+        # Filter by active node ids (alive but not draining)
+        return {
+            node_id: custom_min(gcs_info[node_id], available_minus_replicas[node_id])
```

Comment:
Will we have KeyError?

Reply:
I think not for `total_minus_replicas` because `get_active_node_ids` comes from the same cached dictionary as `total_minus_replicas`. However `gcs_info` makes a separate call to GCS, so I guess there is a slight chance of key error here. Defensively changed it to use an empty resource if `node_id` doesn't exist.